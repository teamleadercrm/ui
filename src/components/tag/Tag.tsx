import { IconCloseSmallOutline } from '@teamleader/ui-icons';
import cx from 'classnames';
import React, { ReactNode, forwardRef } from 'react';
import { GenericComponent } from '../../@types/types';
import { COLORS, SIZES } from '../../constants';
import Box from '../box';
import { BoxProps } from '../box/Box';
import IconButton from '../iconButton';
import { UITextBody, UITextDisplay, UITextSmall } from '../typography';
import theme from './theme.css';

export type TagSize = Exclude<(typeof SIZES)[number], 'tiny' | 'fullscreen' | 'smallest' | 'hero'>;

export interface TagProps extends Omit<BoxProps, 'className' | 'size'> {
  children: ReactNode;
  className?: string;
  onRemoveClick?: React.MouseEventHandler;
  onRemoveMouseDown?: React.MouseEventHandler;
  onRemoveMouseUp?: React.MouseEventHandler;
  removeElement?: React.ElementType;
  size?: TagSize;
  color?: (typeof COLORS)[number];
}

const Tag: GenericComponent<TagProps> = forwardRef<HTMLElement, TagProps>(
  (
    {
      children,
      className,
      onRemoveClick,
      onRemoveMouseDown,
      onRemoveMouseUp,
      removeElement,
      size = 'medium',
      color,
      ...others
    },
    ref,
  ) => {
    const classNames = cx(theme[`is-${size}`], theme['wrapper'], className);

    const TextElement = size === 'small' ? UITextSmall : size === 'large' ? UITextDisplay : UITextBody;

    return (
      <Box {...others} className={classNames} data-teamleader-ui="tag" display="inline-flex" ref={ref}>
        <TextElement
          color={color || 'teal'}
          marginLeft={size === 'small' ? 2 : 3}
          marginRight={onRemoveClick ? 1 : size === 'small' ? 2 : 3}
          marginVertical={size === 'small' ? 0 : 1}
          maxLines={1}
          tint="darkest"
        >
          {children}
        </TextElement>
        {onRemoveClick && (
          <IconButton
            element={removeElement}
            className={theme['remove-button']}
            flexShrink={0}
            icon={<IconCloseSmallOutline />}
            onClick={onRemoveClick}
            onMouseDown={onRemoveMouseDown}
            onMouseUp={onRemoveMouseUp}
            color={color || 'neutral'}
          />
        )}
      </Box>
    );
  },
);

Tag.displayName = 'Tag';

export default Tag;
