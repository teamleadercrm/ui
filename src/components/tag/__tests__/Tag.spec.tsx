import React from 'react';
import { render } from '@testing-library/react';
import Tag from '../Tag';

describe('Component - Tag', () => {
  it('renders', async () => {
    const { asFragment } = render(<Tag>This is a Tag</Tag>);
    expect(asFragment()).toMatchSnapshot();
  });
});
