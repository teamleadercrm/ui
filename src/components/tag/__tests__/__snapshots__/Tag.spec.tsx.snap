// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Tag renders 1`] = `
<DocumentFragment>
  <div
    class="box display-inline-flex is-medium wrapper"
    data-teamleader-ui="tag"
  >
    <p
      class="box margin-bottom-1 margin-left-3 margin-right-3 margin-top-1 ui-text ui-text-body darkest teal overflow-singleline"
      data-teamleader-ui="ui-text"
    >
      This is a Tag
    </p>
  </div>
</DocumentFragment>
`;
